### DISCUSSION TITLE
Tutorial for wsi_dataset

### DISCUSSION THREAD

**Discussion by Dylan-H-Wang** [Project-MONAI/tutorials#1380] [Ideas] [answered]
Hi,

Thank you for open source such an awesome framework for healthcare!

I am currently trying to adapting one of my digital pathology project to MONAI. I have noticed that the [wsi dataset](https://docs.monai.io/en/latest/data.html#whole-slide-image-datasets) and [wsi reader](https://docs.monai.io/en/latest/data.html#whole-slide-image-reader) have been introduced since v0.5, but there is no relevant tutorial to explain how to use them property. It took me a while to understand the source codes and make it work for my cases. Therefore, I am wondering is that possible to write a tutorial to explain how to use them and I believe it would be super helpful?

The second question is how [MaskedPatchWSIDataset](https://github.com/Project-MONAI/MONAI/blob/4e0b179557f5799a4a2a9a32458dfa3afb7eb87f/monai/data/wsi_datasets.py#L321) generates patchs? Since the documentation does not explain it and I checked the [source code](https://github.com/Project-MONAI/MONAI/blob/4e0b179557f5799a4a2a9a32458dfa3afb7eb87f/monai/data/wsi_datasets.py#L405) and have some questions:
1. why `0.5` was chosen as the offset?
2. After minus `patch_size_0 // 2`, it seems like, in the cases of left top part of the image, the mask location will be outside of the boundary.
3. My understanding of the pipeline of this dataset is that it will generate patches based on each of tissue mask pixel. Would not that be too dense? I think it would make more sense to generate non-overlapped patches based on the tissue mask and patch size.

Thank you!

**Comment by KumoLiu** [comment]
Hi @Dylan-H-Wang, thanks for your interest!
Thanks for your suggestion here, adding more tutorials about the `wsi dataset` and `wsi reader` can be helpful. I think we can put it on the agenda, what do you think @drbeh?

And also could you please help share some comments on the questions about `MaskedPatchWSIDataset`, @drbeh?
Thanks in advance!

**Comment by bhashemian** [comment, marked as answer]
Hi @Dylan-H-Wang, thanks for your questions and suggestions.
1. 0.5 is set to make sure that each mask's pixel represent the center of the patch extracted from the image at other levels.
2. That is right for the pixels close to border in the mask, the extracted patch from the image can goes outside of the image.
3. I agree with you the use case that you are explaining is important but not the purpose of `MaskedPatchWSIDataset`. This dataset specifically extracts patches for each and every pixels of the foreground mask and it has it's own use cases.

To understand better your need and to be able to address it in MONAI can you let me know if your use case is for training or inference, and if there is any other feature that you think is missing?

@KumoLiu I agree that we should prepare some tutorials on how to best use pathology datasets.
